cv32e40s/tests/programs/custom/debug_priv_test/debugger.S

```diff
@@ -0,0 +1,225 @@
+	sw s6, 28(sp)
+	sw s7, 32(sp)
+	sw s8, 36(sp)
+	sw s9, 40(sp)
+	sw s10, 44(sp)
+	sw s11, 48(sp)
+
+	la t0, glb_setmprv_test
+	lw t0, 0(t0)
+	bnez t0, MPRV_test
+
+	la t0, glb_check_prv_test
+	lw t0, 0(t0)
+	bnez t0, Check_prv_test
+
+	la t0, glb_check_prv_test_2
+	lw t0, 0(t0)
+	bnez t0, Check_prv_test_2
+
+	la t0, glb_unsupported_mode_test
+	lw t0, 0(t0)
+	bnez t0, Check_unsupported_modes
+
+	la t0, glb_ebreaku_one_test
+	lw t0, 0(t0)
+	bnez t0, Set_ebreaku_bit
+
+	la t0, glb_expect_ebreaku_exc
+	lw t0, 0(t0)
+	bnez t0, Successfull_user_mode
+
+	la t0, glb_check_previous_prv_test
+	lw t0, 0(t0)
+	bnez t0, Check_previous_privilege_test
+
+	j end_debugger
+
+
+end_debugger:
+	// sets the glb_debug_status value high to signal 'debug_priv_test.c' that the debugger is finished executing.
+	la t2, glb_debug_status
+	lw t1, 0(t2)
+	addi t1, t1, 1
+	sw t1, 0(t2)
+	
+	lw ra, 0(sp)
+	lw s0, 4(sp)
+	lw s1, 8(sp)
+	lw s2, 12(sp)
+	lw s3, 16(sp)
+	lw s4, 20(sp)
+	lw s5, 24(sp)
+	lw s6, 28(sp)
+	lw s7, 32(sp)
+	lw s8, 36(sp)
+	lw s9, 40(sp)
+	lw s10, 44(sp)
+	lw s11, 48(sp)
+	addi sp,sp,52        
+    dret
+
+
+
+/*
+Transition out of D-mode (dret) into U-mode, while mstatus.mprv=1, ensure that when execution continues outside D-mode that mstatus.mprv=0.
+*/
+MPRV_test:
+	li t0, 0x3
+	csrrc x0, dcsr, t0 // load U-mode into dcsr.prv register. 
+	li t0, 1
+	slli t0, t0, 17
+	csrrs x0, mstatus, t0 // set MPRV=1
+	j end_debugger
+
+
+/* 
+Transition out of D-mode, ensure that executions starts in the same privilege mode as was indicated in dcsr.prv (dcsr.prv=M and dcsr.prv=U).
+*/
+Check_prv_test:
+	la t0, glb_check_prv_test_val
+	csrrs t1, dcsr, x0 
+	sw t1, 0(t0)
+	j end_debugger
+
+
+/* 
+Transition out of D-mode, ensure that executions starts in the same privilege mode as was indicated in dcsr.prv (dcsr.prv=M and dcsr.prv=U).
+*/
+Check_prv_test_2:
+	li t0, 0x3
+	csrrc x0, dcsr, t0 // load U-mode into dcsr.prv register. 
+
+	la t0, glb_check_prv_test_val
+	csrrs t1, dcsr, x0 
+	sw t1, 0(t0)
+	j end_debugger
+
+
+
+/*
+Write unsupported modes to dcsr.prv, ensure the value read back is unchanged from the previous value.
+*/
+Check_unsupported_modes:
+	li t0, 0x3
+	csrrs x0, dcsr, t0 // preload supported M-mode to get a reference
+
+	// Check S-mode
+	li t1, 0x1
+	csrrs x0, dcsr, t1
+	csrrs t1, dcsr, x0
+	and t1, t1, t0
+	bne t1, t0, Unsupported_check_failed
+
+	// Check R-mode 
+	li t1, 0x2
+	csrrs x0, dcsr, t1
+	csrrs t1, dcsr, x0
+	and t1, t1, t0
+	bne t1, t0, Unsupported_check_failed
+
+	j end_debugger
+
+
+/*
+Write unsupported modes to dcsr.prv, ensure the value read back is unchanged from the previous value.
+*/
+Unsupported_check_failed: // if unsupported mode is written to dcsr.prv go here 
+	la t0, glb_unsupported_check_failed
+	li t1, 1
+	sw t1, 0(t0)
+	j end_debugger
+
+
+
+/*
+Have dcsr.ebreaku=1, be in U-mode, execute ebreak, ensure debug entry happens instead of "normal" ebreak behavior.
+*/
+Set_ebreaku_bit:
+	lw t0, 0(x0)
+	lui t0, 1
```

Comment:
While I like to avoid pseudo instructions, I think `li` would be easier to read here. Because I can't quickly determine if the correct bit gets set.

Reply:
I think It would make sense to add a li with some #define value specifying what the value is. I can do that

Reply:
That would be nice.